src/to-be-required.js

```diff
@@ -0,0 +1,68 @@
+import {matcherHint, printReceived} from 'jest-matcher-utils'
+import {checkHtmlElement} from './utils'
+
+// form elements that support 'required'
+const FORM_TAGS = [
+  'input',
+  'select',
+  'optgroup',
+  'option',
+  'button',
+  'textarea',
+]
+
+function getTag(element) {
+  return element.tagName && element.tagName.toLowerCase()
+}
+
+function isElementRequired(element) {
+  return FORM_TAGS.includes(getTag(element)) && element.hasAttribute('required')
+}
+
+function isElementRequiredByARIA(element) {
+  return (
+    FORM_TAGS.includes(getTag(element)) &&
```

Comment:
On the other hand I suspect that `aria-selected` is not as restricted, and may be usable even in `div` elements with certain `role`s ([ref](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-required_attribute#Used_in_ARIA_roles)). Though again, I'm not suggesting we go as far as ensuring that if present the element has one of these roles (unless you think is easily doable and makes sense). But at least we should probably not enforce that the element tag is one of these.

Reply:
We could enforce that the attribute is present on either:

* a block having `role="combobox"`, `role="gridcell"`, `role="radiogroup"`, `role="spinbutton"` or `role="tree"`
* a block having or a form element corresponding to the `role="listbox"` (so any `<selectbox>`)
* an block having or a form element corresponding to the `role="textbox"` (so any `<input>` and `<textarea>`)